### ISSUE TITLE
1.18.1 JEI Issues with this mod on servers

### ISSUE THREAD

**Issue by Charmanth** [cech12/BucketLib#8]
So I'm running a 1.18.1 server, and found an issue with bucketlib that causes JEI to not work properly.  The recipes won't load from the server to the client.

[Logs](https://gist.github.com/Charmanth/b95a4746a455938e0bc1f023f635d4ab)

[JEI Issue for reference](https://github.com/mezz/JustEnoughItems/issues/2749)

**Comment by mezz** [comment]
(Unfortunately) this is not just affecting JEI, it is breaking recipe sync from the server to the client.

```
[29Mar2022 18:40:46.748] [Render thread/FATAL] [net.minecraft.util.thread.BlockableEventLoop/]: Error executing task on Client
net.minecraft.ResourceLocationException: Non [a-z0-9/._-] character in path of location: minecraft:net.minecraft.world.level.material.WaterFluid$Source@4bf89d3d
	at net.minecraft.resources.ResourceLocation.<init>(ResourceLocation.java:37) ~[client-1.18.1-20211210.034407-srg.jar%23182!/:?]
	at net.minecraft.resources.ResourceLocation.<init>(ResourceLocation.java:42) ~[client-1.18.1-20211210.034407-srg.jar%23182!/:?]
	at cech12.bucketlib.api.crafting.FluidIngredient$Serializer.parse(FluidIngredient.java:153) ~[bucketlib-1.18.1-0.5.0.1.jar%2369!/:1.18.1-0.5.0.1]
	at cech12.bucketlib.api.crafting.FluidIngredient$Serializer.parse(FluidIngredient.java:138) ~[bucketlib-1.18.1-0.5.0.1.jar%2369!/:1.18.1-0.5.0.1]
	at net.minecraftforge.common.crafting.CraftingHelper.getIngredient(CraftingHelper.java:95) ~[forge-1.18.1-39.1.2-universal.jar%23186!/:?]
	at net.minecraft.world.item.crafting.Ingredient.m_43940_(Ingredient.java:193) ~[client-1.18.1-20211210.034407-srg.jar%23182!/:?]
	at net.minecraft.world.item.crafting.ShapelessRecipe$Serializer.m_8005_(ShapelessRecipe.java:111) ~[client-1.18.1-20211210.034407-srg.jar%23182!/:?]
	at net.minecraft.world.item.crafting.ShapelessRecipe$Serializer.m_8005_(ShapelessRecipe.java:77) ~[client-1.18.1-20211210.034407-srg.jar%23182!/:?]
	at net.minecraft.network.protocol.game.ClientboundUpdateRecipesPacket.m_133647_(ClientboundUpdateRecipesPacket.java:45) ~[client-1.18.1-20211210.034407-srg.jar%23182!/:?]
	at net.minecraft.network.FriendlyByteBuf.m_178371_(FriendlyByteBuf.java:119) ~[client-1.18.1-20211210.034407-srg.jar%23182!/:?]
	at net.minecraft.network.FriendlyByteBuf.m_178366_(FriendlyByteBuf.java:135) ~[client-1.18.1-20211210.034407-srg.jar%23182!/:?]
	at net.minecraft.network.protocol.game.ClientboundUpdateRecipesPacket.<init>(ClientboundUpdateRecipesPacket.java:22) ~[client-1.18.1-20211210.034407-srg.jar%23182!/:?]
	at net.minecraft.network.ConnectionProtocol$PacketSet.m_178327_(ConnectionProtocol.java:417) ~[client-1.18.1-20211210.034407-srg.jar%23182!/:?]
	at net.minecraft.network.ConnectionProtocol.m_178321_(ConnectionProtocol.java:469) ~[client-1.18.1-20211210.034407-srg.jar%23182!/:?]
	at net.minecraftforge.network.filters.VanillaPacketSplitter.onClientPacket(VanillaPacketSplitter.java:146) ~[forge-1.18.1-39.1.2-universal.jar%23186!/:?]
	at net.minecraftforge.eventbus.EventBus.doCastFilter(EventBus.java:247) ~[eventbus-5.0.3.jar%2337!/:?]
	at net.minecraftforge.eventbus.EventBus.lambda$addListener$11(EventBus.java:239) ~[eventbus-5.0.3.jar%2337!/:?]
	at net.minecraftforge.eventbus.EventBus.post(EventBus.java:302) ~[eventbus-5.0.3.jar%2337!/:?]
	at net.minecraftforge.eventbus.EventBus.post(EventBus.java:283) ~[eventbus-5.0.3.jar%2337!/:?]
	at net.minecraftforge.network.NetworkInstance.dispatch(NetworkInstance.java:68) ~[forge-1.18.1-39.1.2-universal.jar%23186!/:?]
	at net.minecraftforge.network.NetworkHooks.lambda$onCustomPayload$1(NetworkHooks.java:75) ~[forge-1.18.1-39.1.2-universal.jar%23186!/:?]
	at java.util.Optional.map(Optional.java:260) ~[?:?]
	at net.minecraftforge.network.NetworkHooks.onCustomPayload(NetworkHooks.java:75) ~[forge-1.18.1-39.1.2-universal.jar%23186!/:?]
	at net.minecraft.client.multiplayer.ClientPacketListener.m_7413_(ClientPacketListener.java:1812) ~[client-1.18.1-20211210.034407-srg.jar%23182!/:?]
	at net.minecraft.network.protocol.game.ClientboundCustomPayloadPacket.m_5797_(ClientboundCustomPayloadPacket.java:57) ~[client-1.18.1-20211210.034407-srg.jar%23182!/:?]
	at net.minecraft.network.protocol.game.ClientboundCustomPayloadPacket.m_5797_(ClientboundCustomPayloadPacket.java:7) ~[client-1.18.1-20211210.034407-srg.jar%23182!/:?]
	at net.minecraft.network.protocol.PacketUtils.m_131356_(PacketUtils.java:21) ~[client-1.18.1-20211210.034407-srg.jar%23182!/:?]
	at net.minecraft.util.thread.BlockableEventLoop.m_6367_(BlockableEventLoop.java:151) ~[client-1.18.1-20211210.034407-srg.jar%23182!/:?]
	at net.minecraft.util.thread.ReentrantBlockableEventLoop.m_6367_(ReentrantBlockableEventLoop.java:23) ~[client-1.18.1-20211210.034407-srg.jar%23182!/:?]
	at net.minecraft.util.thread.BlockableEventLoop.m_7245_(BlockableEventLoop.java:125) ~[client-1.18.1-20211210.034407-srg.jar%23182!/:?]
	at net.minecraft.util.thread.BlockableEventLoop.m_18699_(BlockableEventLoop.java:110) ~[client-1.18.1-20211210.034407-srg.jar%23182!/:?]
	at net.minecraft.client.Minecraft.m_91383_(Minecraft.java:1010) ~[client-1.18.1-20211210.034407-srg.jar%23182!/:?]
	at net.minecraft.client.Minecraft.m_91374_(Minecraft.java:661) ~[client-1.18.1-20211210.034407-srg.jar%23182!/:?]
	at net.minecraft.client.main.Main.main(Main.java:205) ~[client-1.18.1-20211210.034407-srg.jar%23182!/:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:568) ~[?:?]
	at net.minecraftforge.fml.loading.targets.CommonClientLaunchHandler.lambda$launchService$0(CommonClientLaunchHandler.java:45) ~[fmlloader-1.18.1-39.1.2.jar%2351!/:?]
	at cpw.mods.modlauncher.LaunchServiceHandlerDecorator.launch(LaunchServiceHandlerDecorator.java:37) [modlauncher-9.1.3.jar%2340!/:?]
	at cpw.mods.modlauncher.LaunchServiceHandler.launch(LaunchServiceHandler.java:53) [modlauncher-9.1.3.jar%2340!/:?]
	at cpw.mods.modlauncher.LaunchServiceHandler.launch(LaunchServiceHandler.java:71) [modlauncher-9.1.3.jar%2340!/:?]
	at cpw.mods.modlauncher.Launcher.run(Launcher.java:106) [modlauncher-9.1.3.jar%2340!/:?]
	at cpw.mods.modlauncher.Launcher.main(Launcher.java:77) [modlauncher-9.1.3.jar%2340!/:?]
	at cpw.mods.modlauncher.BootstrapLaunchConsumer.accept(BootstrapLaunchConsumer.java:26) [modlauncher-9.1.3.jar%2340!/:?]
	at cpw.mods.modlauncher.BootstrapLaunchConsumer.accept(BootstrapLaunchConsumer.java:23) [modlauncher-9.1.3.jar%2340!/:?]
	at cpw.mods.bootstraplauncher.BootstrapLauncher.main(BootstrapLauncher.java:149) [bootstraplauncher-1.0.0.jar:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:568) ~[?:?]
	at io.github.zekerzhayard.forgewrapper.installer.Main.main(Main.java:57) [ForgeWrapper-mmc3.jar:mmc3]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:568) ~[?:?]
	at org.multimc.onesix.OneSixLauncher.launchWithMainClass(OneSixLauncher.java:210) [NewLaunch.jar:?]
	at org.multimc.onesix.OneSixLauncher.launch(OneSixLauncher.java:245) [NewLaunch.jar:?]
	at org.multimc.EntryPoint.listen(EntryPoint.java:143) [NewLaunch.jar:?]
	at org.multimc.EntryPoint.main(EntryPoint.java:34) [NewLaunch.jar:?]
```

Since this is a library, the best "fix" might be to validate the input as early as possible when it's passed into the library, and make it crash earlier. That will let us identify what mod the problem is coming from and prevent creating this type of invalid state down the line.

**Comment by cech12** [comment]
Thanks for your report! I will look into that as soon as possible. :)

**Comment by cech12** [comment]
@Charmanth 
I released a new version [1.18.1-0.5.0.2](https://github.com/cech12/BucketLib/releases/tag/1.18.1-0.5.0.2)
Please test it and let me know if it works. :)

**Comment by Charmanth** [comment]
That seems to fix it!
